### PR TITLE
[xla:gpu] Fine grained PJRT definition events for result buffers

### PR DESCRIPTION
**Pull request by ezhulenev**
StreamExecutor PJRT currently gives every result buffer the same definition event, recorded after the entire executable has been scheduled. This unnecessarily delays consumers of results whose final device work was scheduled earlier.

This change integrates XLA:GPU’s ThunkExecutor definition tracking with PJRT (built in top of https://github.com/openxla/xla/pull/46914):

  - Extends `RawExecuteResult` with optional per-result definition events. Results without an individual event fall back to the executable-wide event.
  - Creates a definition-event promise for every result before asynchronous dispatch.
  - Maps GPU buffer allocations to their corresponding PJRT result leaves.
  - Installs a `ThunkExecutor` definition callback during execution. The callback records one event on the stream that scheduled the allocation’s final use and resolves every result sharing that allocation.
  - Resolves any promises not handled by the definition callback with the executable-wide event, preserving existing behavior for untracked allocations and failures.
  - Updates `CommonPjRtClient::CreateOutputs` to attach each result’s individual definition event.
  - Materializes the allocation-readiness sync point before launch. Allocation readiness remains separate from result-content definition.

A lot of churn is because of previous commits that didn't run clang-format when moved to `ABSL_ASSIGN_OR_RETURN`.

### CODE REVIEW

**Comment by pschuh** [comment]
I feel like this should be more optional than it is (maybe a mask of the buffers where you want to use something other than the primary execute event that we pass into execute options?).

**Comment by ezhulenev** [comment]
Yeah, it might be too broad, the real benefit will show up only in a few limited use cases (where user does not run on the same compute stream).

How about a JAX API like this:

```
  with jax.experimental.output_definition_events([0, 3, 4]):
    results = executable(lhs, rhs)
``` 

With this semantics:
  - Indices refer to flattened executable outputs.
  - Only executions launched in the current thread and scope are affected.
  - Unselected outputs use the primary execute event.
  - Selected outputs may still conservatively fall back to the primary event.

@Angelogeb I think for the jaxpp case it should work really well, outputs that cross the pipeline boundary will have to be explicitly marked to get their own definition event (jaxpp already knows this), and everything else gets the default event.

**Comment by pschuh** [comment]
I was thinking about it more and because there is some expense at runtime for allocating events, we probably want to compile into the executable some sort of notion of groups of buffers that get finalized together. What is the current strategy for building the groups of indices? We could probably also attach output attributes on the mlir module and pick them up during compilation without too much hassle.

**Comment by Angelogeb** [comment]
> @Angelogeb I think for the jaxpp case it should work really well, outputs that cross the pipeline boundary will have to be explicitly marked to get their own definition event (jaxpp already knows this), and everything else gets the default event.

Yes, I think making this opt in by marking which outputs should get their own definition event is sensible.

> We could probably also attach output attributes on the mlir module and pick them up during compilation without too much hassle

Having `jax.jit(..., compiler_options={'output_definition_events': [0, 3, 4]})` can work too, instead of having a context manager at the call site.

**Comment by ezhulenev** [comment]
I don't like `compiler_options=...` because it's not really a property of XLA executable, and not a compile-time flag, the same jit executable can be executed with different masks for output definition events, it's not even XLA:GPU runtime property, it stays in SE/PJRT client.

I can't think of any prior art in JAX/XLA that falls into the same category.

**Comment by ezhulenev** [comment]
`What is the current strategy for building the groups of indices?`

Today `ThunkExecutor` simply infers the last thunk that touches buffer allocation with given index, and this is essentially inferred from buffer assignment (last HLO that reads/writes from buffer allocation corresponding to output).

This inferred once and cached in GpuExecutable, and the cost of callback is ~0, so only PJRT client has to make a decision if it wants a fine grained event for a particular buffer (which is not free) or reuse a coarse grained event for execution.

**Comment by ezhulenev** [comment]
@pschuh how about this approach, I added `individually_defined_output_indices` which follows prior art of `non_donatable_input_indices`?

File: xla/pjrt/se/pjrt_stream_executor_client.cc [hunk 1]
```diff
@@ -1549,21 +1554,37 @@ PjRtStreamExecutorRawLoadedExecutable::Execute(
     gpu_run_options->set_incarnations(std::move(device_incarnations));
   }
 
+  std::vector<PjRtDeviceEventPromiseRef> result_definition_event_promises(
+      results.size());
+  std::vector<PjRtDeviceEventRef> result_definition_events(results.size());
+  for (int result_index : options.individually_defined_output_indices) {
```
Comment (pschuh) [inline xla/pjrt/se/pjrt_stream_executor_client.cc]: We should completely disable this feature if individually_defined_output_indices.size() == 0.

Comment (ezhulenev) [reply xla/pjrt/se/pjrt_stream_executor_client.cc]: Done, pass empty spans everywhere when executable uses default event.

**Review by pschuh — APPROVED** [review]

**Comment by hyeontaek** [comment]
I think CompileOptions is a better place for this feature. In theory, the compiler can emit a program that computes some outputs "as soon as possible" if it is asked to declare the definition of these individual outputs, and this type of handling would be in line with input/output aliasing and donated input handling. Furthermore, it is hard to imagine frequently changing the `output_definition_events` at each execute, and being able to change it at each Execute will be an unnecessary runtime overhead.

IFRT is neutral of when outputs should become ready (e.g., IFRT IR programs already make some outputs ready as soon as the atom program computes them finishes). In other words, `individually_defined_output_indices` is not well defined at the IFRT level, and this concept should be rather handled directly in PjRt or lower compiler/runtime layers, which can be done with CompileOptions more easily today.

**Comment by pschuh** [comment]
I think hyeontaek makes some good points.

**Comment by ezhulenev** [comment]
Agree. Pushed a new version that added a new flag to compilation options.

File: xla/pjrt/gpu/se_gpu_pjrt_client.cc [hunk 2]
```diff
@@ -1155,16 +1158,16 @@ StreamExecutorGpuRawClient::CrossHostReceiveBuffersInto(
...
+          &flag_address, 0, sizeof(uint32_t) * raw_buffers.size()));
 
       // Export the address range which contains the flag buffer. The flags are
       // addressed as offsets from the beginning of this range.
       ABSL_ASSIGN_OR_RETURN(
           const std::string flag_handle,
           GetOrExportFabricHandle(executor, flag_address.opaque()));
-      ABSL_ASSIGN_OR_RETURN(auto flag_range,
-                       executor->GetAllocationRange(flag_address.opaque()));
+      ABSL_ASSIGN_OR_RETURN(
```
Comment (pschuh) [inline xla/pjrt/gpu/se_gpu_pjrt_client.cc]: The ASSIGN_OR_RETURN -> ABSL_ASSIGN_OR_RETURN was implemented as a "rewrite during import/export process":
https://github.com/openxla/xla/commit/eb278d3e56cac0341e3a9cfaa6658ae79c2808f4

so if you could focus the clang-format to just the updated lines that would be helpful for the review. Also, just FYI, it appears that the import process runs the auto-formatting for you so it will all get erased when you merge.

Comment (ezhulenev) [reply xla/pjrt/gpu/se_gpu_pjrt_client.cc]: Hmmm, the problem is that it will break OSS clang-format presubmit as it verifies format for all changed files, not only changed lines. It has to be somehow fixed on google side.

I uploaded new diff without unrelated formatting changes, maybe it will get past internal Google CI. But in general this "rewrite during import/export process" will inevitably create a lot of churn because clang-format formats everything.

I'd say that similar to import process, export should do clang-format based on OSS config.

Comment (ezhulenev) [reply xla/pjrt/gpu/se_gpu_pjrt_client.cc]: Ok, looks like clang-format check actually is smart enough to check only changed lines, so the only issue is with extra pain for OSS development where clang format formats whole file by default.

File: xla/pjrt/pjrt_executable.h [hunk 3]
```diff
@@ -91,6 +92,10 @@ struct CompileOptions {
   // tuple and passed as a single parameter.
   bool parameter_is_tupled_arguments = false;
 
+  // Flattened output indices that should use individual definition events when
+  // supported. Other outputs use the primary execute event.
+  absl::btree_set<int> individually_defined_output_indices;
```
Comment (pschuh) [inline xla/pjrt/pjrt_executable.h]: Wait, why btree_set? Isn't it just as efficient but simpler to require them to be a sorted vector (test using std::lower_bound) ?

File: xla/pjrt/se/pjrt_stream_executor_client.h [hunk 4]
```diff
@@ -398,9 +412,12 @@ class PjRtStreamExecutorRawClient : public PjRtRawClient {
 
 class PjRtStreamExecutorExecutableLoadState : public PjRtExecutableLoadState {
  public:
-  explicit PjRtStreamExecutorExecutableLoadState(
-      PjRtStreamExecutorRawClient* raw_client)
-      : raw_client_(raw_client) {}
+  PjRtStreamExecutorExecutableLoadState(
+      PjRtStreamExecutorRawClient* raw_client,
+      std::vector<int> individually_defined_output_indices)
```
Comment (pschuh) [inline xla/pjrt/se/pjrt_stream_executor_client.h]: Wait, why is this here? Isn't this available on the CompileOptions that we can get from the PjRtExecutable?These days we serialize both the CompileOptions and the platform device executable into the serialized blob for GPU so that should be the source of truth.

File: xla/pjrt/pjrt_executable.h (hunk 3 above)
Comment (ezhulenev) [reply xla/pjrt/pjrt_executable.h]: Just to avoid sorting when serializing options to proto, which is needed for determinism, `flat_hash_set` iteration is random. Basically either btree set here, or sorting when iteration order must be well defined.

Comment (ezhulenev) [reply xla/pjrt/pjrt_executable.h]: Hm, `non_donatable_input_indices` actually don't even bother with deterministic serialization in `ExecuteOptions::ToProto`, but I'm not sure if it's important for execution options, compilation options AFAIK are a part of the compilation cache key.

Anyway, decided to keep flat hash set for consistency and sort during ToProto serialization.

File: xla/pjrt/se/pjrt_stream_executor_client.h (hunk 4 above)
Comment (ezhulenev) [reply xla/pjrt/se/pjrt_stream_executor_client.h]: Changed to capturing `CompileOptions` by reference. It seems to be safe today, but does loaded executable requires `PjRtExecutable` to be alive?

**Review by pschuh — APPROVED** [review]

File: xla/pjrt/se/pjrt_stream_executor_client.h (hunk 4 above)
Comment (pschuh) [reply xla/pjrt/se/pjrt_stream_executor_client.h]: To be safe, PjRtStreamExecutorRawLoadedExecutable should just hold a ref to the AsyncValueRef<PjRtExecutable>?

Comment (ezhulenev) [reply xla/pjrt/se/pjrt_stream_executor_client.h]: Yeah, we already have a ref anyway. Added a new member.

**Review by pschuh — APPROVED** [review]

**Review by hyeontaek — APPROVED** [review]

**Review by pschuh — APPROVED** [review]

File: xla/pjrt/gpu/BUILD [hunk 5]
```diff
@@ -100,6 +100,7 @@ cc_library(
         "//xla/backends/gpu/collectives:gpu_cliques",
         "//xla/backends/gpu/collectives:gpu_collectives",
         "//xla/backends/gpu/collectives:gpu_communicator",
+        "//xla/backends/gpu/runtime:thunk_executor",
```
Comment (pschuh) [inline xla/pjrt/gpu/BUILD]: This dep has compat build problems internally. I think it needs to be in if_cuda_or_rocm and you need to #if guard on it.

Comment (ezhulenev) [reply xla/pjrt/gpu/BUILD]: Added it to two conditional sections

**Review by pschuh — APPROVED** [review]
